### ISSUE TITLE
Dev image does not send proper url for webhook callbacks

### ISSUE THREAD

**Issue by naanlizard** [AirenSoft/OvenMediaEngine#1739]
**Describe the bug**
On the latest dev image (`airensoft/ovenmediaengine@sha256:773acc79230548f8d942b2321f94d10e0846135128a5b0dc35eab60ff0fa528e`), admission webhooks do not work because the URL sent does not include the streamkey

**To Reproduce**
[Server.xml](https://gist.github.com/naanlizard/8c1ce4fe2ae9f09bda21a17f69a09f63)

(Note, in Server.xml, the Decodes section is either commented out (for 0.17.1) or uncommented (for dev)

[Dockerfile](https://gist.github.com/naanlizard/73927d1e198ca8f48e585064e4eec4bb)

Maybe relevant, if you put the streamkey in the URL (e.g. in OBS, your rtmp url is set to `rtmp://ingest.hostname.tld:1935/live/streamkey`) both versions work. 

**Expected behavior**
URL sent should include the streamkey, like on 0.17.1.

**Logs**
Relevant log lines from our backend API, Parameters being the values from the request OME sends

0.17.1
`
Parameters: {"client"=>{"address"=>"x", "port"=>58323, "real_ip"=>"x"}, "request"=>{"direction"=>"incoming", "protocol"=>"RTMP", "status"=>"opening", "time"=>"2024-11-14T15:23:54.039+00:00", "url"=>"rtmp://ingest.hostname.tld:1935/live/streamkey"}, "stream"=>{}}
`

dev
`Parameters: {"client"=>{"address"=>"x", "port"=>58619, "real_ip"=>"x"}, "request"=>{"direction"=>"incoming", "protocol"=>"RTMP", "status"=>"opening", "time"=>"2024-11-14T15:28:11.658+00:00", "url"=>"rtmp://ingest.hostname.tld:1935/live"}, "stream"=>{}}`

Additional info:

~~Interestingly, it seems like the closing webhook *does* send the streamkey in the URL~~

~~`Request params: {"direction"=>"incoming", "new_url"=>"rtmp://ingest.hostname.tld:1935/live/streamkey", "protocol"=>"RTMP", "status"=>"closing", "time"=>"2024-11-14T15:27:15.762+00:00", "url"=>"rtmp://apollo-rtmp://ingest.hostname.tld:1935/live/streamkey"}`~~

Double checked the above, it seems like it does NOT send the streamkey in the closing call, or at least not always. Strange

**Comment by dimiden** [comment]
@naanlizard
Thank you for the report! 
Upon reviewing the issue you mentioned, I found that a recent improvement in the URL parsing process caused the problem. 
I have just fixed this issue and committed it as 9f92959b946e8fd632c56ae87d803b8d78b5375f.

**Comment by naanlizard** [comment]
The proper URL is now sent, but there's a new problem, which I'm not sure how to describe

[omebug111624.log-clean.txt](https://github.com/user-attachments/files/17787358/omebug111624.log-clean.txt)

Dockerfile is the same as above but with 

`FROM airensoft/ovenmediaengine@sha256:0510bb6178a96083139a66b601c8aceeba5bb787f2e5694c8abfbcbb0010edef`

**Comment by dimiden** [comment]
@naanlizard 
Oh no... The logs show there’s an issue. I thought the URL parser was modified to be compatible with the previous code, but it seems there are unexpected side effects. 
I will carefully check if there are any other issues besides the one you mentioned and get back to you!

**Comment by dimiden** [comment]
@naanlizard 
The issue occurred because a different value was returned for invalid URL, such as empty URL, compared to the previous implementation. 
This issue has been fixed in 4c4ca98ab14a37080292f8792952124eaaca591e, and it has been confirmed that it works correctly even if empty URL is responsed from admission webhook server.

Thank you for the testing!

**Comment by naanlizard** [comment]
Thanks! I'll test it asap

Can you clarify what the admission server should return for that url value?
Are we doing something unusual?

**Comment by naanlizard** [comment]
Hmm, I think we set a new_url, but I will double check and get back to you

**Comment by dimiden** [comment]
Yes. If `new_url` is not empty, the issue might be with the regular expression used for URL parsing. It would be helpful if you could provide the URL being passed as `new_url` by the server.
(Since the URL pattern is important, you may replace sensitive information with arbitrary characters.)

For reference, I tested the parser with various URL combinations and encountered no issues, so I suspect there’s a high probability that a empty string was responsed.

**Comment by naanlizard** [comment]
It could be a bug in our backend code, I'll verify the url first 👍

**Comment by hernanrz** [comment]
Looks like the latest dev docker image fixed this, we were responding to OME admission webhook with the following response, which wasn't working on image hash `sha256:0510bb6178a96083139a66b601c8aceeba5bb787f2e5694c8abfbcbb0010edef`

```
{"allowed":true,"new_url":"live/1"}
```

After further testing on that image I found that if you set the url to an absolute URL, it worked again

![imagen](https://github.com/user-attachments/assets/36bbf8c9-8384-4495-a313-c70cc5e9bab5)

After pulling the latest dev docker image `sha256:103e6e0efafe507c6e6f3065d457a1fd2db11ce87a2a74539db8da963716a70a` it works again with the relative URL format

**Comment by naanlizard** [comment]
See hernan's comment above - what is the intended use case for new_url? 

For our purpose, we have users connect to rtmp://hostname.tld:1935/live and put userid?streamkey in their streamkey slot in OBS. So the URL OME sees is in sum, rtmp://hostname.tld:1935/live/userid?streamkey

Should we be returning the /live/userid value, so that the playback url is correct (e.g. https://hostname.tld/live/1/llhls.m3u8) or is that unnecessary?

If so, should new_url be the full URL (e.g. rtmp://hostname.tld:1935/live/userid) or just the relative path (e.g. /live/userid)

I'll make a pull request for documentation improvements once I understand

**Comment by choigilhoon** [comment]
@naanlizard 

First, the `new_url` should be set in the form of a full URL (`scheme://host[:port]/app/stream/file?query=value&query2=value2`). (Please refer to [this](https://airensoft.gitbook.io/ovenmediaengine/dev/access-control/admission-webhooks#format-1))
Then, depending on the form of the playback URL, the value of new_url should change.

Below are two use cases for setting `new_url`:
[Use case 1]
If you want to hide the streamkey and expose only the userid in the playback URL, such as https://hostname.tld/live/{userid}/llhls.m3u8, set `new_url` as follows:
**"rtmp://hostname.tld:1935/live/{userid}"**

[Use case 2]
If you want to hide the userid and expose only the streamkey in the playback URL, such as https://hostname.tld/live/{streamkey}/llhls.m3u8, set `new_url` as follows:
**"rtmp://hostname.tld:1935/live/{streamkey}"**


And, if `new_url` is not a full URL, this value will be ignored, and the value set in OBS (`rtmp://hostname.tld:1935/live/userid?streamkey`) will be used.

This explains why, even though you entered a relative path in `new_url`, it seems to work properly because the stream name is automatically set to {userid}, as in **[Use case 1]**.

**Comment by naanlizard** [comment]
To be clear @choigilhoon -

In a vanilla OME install/config, sending an OBS url and streamkey like

`rtmp://hostname.tld:1935/appname`

and 

`abc?xyz123` (or any random strings separated by a `?`)

Will lead to OME serving that stream as

`https://hostname.tld/appname/abc/llhls.m3u8`

Correct?

**Comment by choigilhoon** [comment]
@naanlizard 

That's correct.

In the situation you mentioned, you should respond to the OME admission webhook with one of the cases below.

* Case 1
```
{
  "allowed": true
}
```
* Case 2
```
{
  "allowed": true,
  "new_url": "rtmp://hostname.tld:1935/appname/abc"
}
```

> Please note that the query-string part of the OBS streamkey (`?xyz123`) is ignored when generating the playback URL.

**Comment by Byorun** [comment]
I just updated to the latest version from pre 17.X and also noticed that my admission webhooks did not work and even if version 17.2. has this line "Fixed an issue where AdmissionWebhooks might not include the stream name" it is not working for me.

Prior to 17.2. I could connect with obs via rtmp://hostname/live/username (stream name) and providing the streamkey via the designated field of obs. after 17.1 I cannot login because the url is missing the username part in the url.

```
<Applications>
    <Application>
        <Name>live</Name>
        <Type>live</Type>
...
```
I only configured the application "live" and since the "username" or streamname was included in the admission webhook I use this regex `^(https?|rtmps?):\/\/.*(:[0-9]+)?\/(live\/[a-z0-9]+)(\/(.*)|-.*)$` to get alle the infos I needed from `rtmp://stream-hostname:1935/live/username/my-secret-streamkey`

**Comment by dimiden** [comment]
@Byorun
I’m a bit confused, so I’d like to ask for clarification.  
Are you saying that you entered `rtmp://hostname/live/username` in the `Server` field and `my-secret-streamkey` in the `Stream Key` field in OBS?

**Comment by Byorun** [comment]
@dimiden yes exactly
